src/main/php/lang/ast/Emitter.class.php

```diff
@@ -310,43 +313,75 @@ protected function emitSignature($signature) {
   }
 
   protected function emitFunction($function) {
+    $this->stack[]= $this->locals;
+    $this->locals= [];
+
     $this->out->write('function '.$function->name); 
     $this->emitSignature($function->signature);
 
     $this->out->write('{');
     $this->emit($function->body);
     $this->out->write('}');
+
+    $this->locals= array_pop($this->stack);
   }
 
   protected function emitClosure($closure) {
+    $this->stack[]= $this->locals;
+    $this->locals= [];
+
     $this->out->write('function'); 
     $this->emitSignature($closure->signature);
 
     if ($closure->use) {
       $this->out->write(' use('.implode(',', $closure->use).') ');
+      foreach ($closure->use as $name) {
+        $this->locals[substr($name, 1)]= true;
```

Comment:
It would be great if this were consistent with other parts, where the names come without the leading `$`.

Reply:
Changing this would break *use by reference*, which is created inside the `Parse` class as follows:

```php
if ('&' === $this->token->value) {
  $this->token= $this->advance();
  $use[]= '&$'.$this->token->value;
} else {
  $use[]= '$'.$this->token->value;
}
```

We could create structs with `byref` and `name` members, and then emit that with some conditionals like inside `emitParameter()`; but that would make this unnecessarily complex.